manifests/install.pp

```diff
@@ -18,17 +18,43 @@
     if $::operatingsystem != 'darwin' {
       ensure_packages(['unzip'])
     }
-    staging::file { 'consul.zip':
-      source => $consul::download_url
-    } ->
-    staging::extract { 'consul.zip':
-      target  => $consul::bin_dir,
-      creates => "${consul::bin_dir}/consul",
-    } ->
-    file { "${consul::bin_dir}/consul":
-      owner => 'root',
-      group => 0, # 0 instead of root because OS X uses "wheel".
-      mode  => '0555',
+
+    # Installing consul for the very first time
+    if ($::consul_version == undef) {
+
+      staging::file { "consul_${consul::version}.zip":
+        source => $consul::download_url,
+      } ->
+      staging::extract { "consul_${consul::version}.zip":
+        target => $consul::bin_dir,
+      } ->
+      file { "${consul::bin_dir}/consul":
+        owner => 'root',
+        group => 0, # 0 instead of root because OS X uses "wheel".
+        mode  => '0555',
+      }
+
+    # Upgrading existing installation of consul
+    } elsif ($::consul_version != $consul::version) {
+
+      exec { "removing consul v${::consul_version}":
```

Comment:
Thats fine, although I would rather have the entire install class notify the entire run_service class per this pattern
https://www.devco.net/archives/2012/12/13/simple-puppet-module-structure-redux.php